api/processors/services/group_builder.py

```diff
@@ -146,23 +150,120 @@ def _get_allowed_group_size(age_range: AgeRange) -> GroupSizeRestriction:
         raise ValueError(f"Group age range is inconsistent with boundaries: {age_range}")
 
     @staticmethod
-    def _get_group_candidate(teacher_id: int) -> GroupCandidate | None:
-        # THIS IS A STUB METHOD RETURNING DUMMY VALUES
-        # TODO filter for language+level, communication language, age groups, time slots (all UTC?)
-        # TODO iterate in correct priority order, yield results
+    def _get_best_group_candidate(teacher_id: int) -> GroupCandidate | None:
         teacher = Teacher.objects.get(pk=teacher_id)
+        group_candidates: list[GroupCandidate] = []
+
+        for language_and_level in teacher.teaching_languages_and_levels.all():
+            logger.info(language_and_level)
+            for first_time_slot, second_time_slot in GroupBuilder._iterate_lesson_times(teacher):
+                for age_range in teacher.student_age_ranges.all():
+                    group_candidate = GroupBuilder._get_group_candidate(
+                        teacher=teacher,
+                        age_range=age_range,
+                        language_and_level=language_and_level,
+                        day_time_slots=(first_time_slot, second_time_slot),
+                    )
+                    if group_candidate is not None:
+                        group_candidates.append(group_candidate)
+
+        if len(group_candidates) == 0:
+            return None
+
+        group_candidates.sort(key=lambda candidate: candidate.priority, reverse=True)
+        return group_candidates[0]
```

Comment:
```suggestion
        return sorted(group_candidates, key=lambda candidate: candidate.priority).pop() if group_candidates else None
```

Reply:
(nit: `group_candidates` is created within the function, so it's not crucial to use `sorted()` instead of `.sort`)

Reply:
Yes, sorry, here my personal opinion is about pop() being less readable in a one'liner. I'll stick to it here. It's a neverending discussion about "less lines means more readable or not?" and in many cases there's no true right and wrong :)